src/yb/yql/cql/cqlserver/cql_service.cc

```diff
@@ -229,6 +234,29 @@ void CQLServiceImpl::DeletePreparedStatement(const shared_ptr<const CQLStatement
           << ", memory usage = " << prepared_stmts_mem_tracker_->consumption();
 }
 
+bool CQLServiceImpl::CheckPassword(const std::string plain, const std::string hash) {
+  // FIXME DO NOT MERGE: hash the plaintext password so we're not keeping
+  // multiple plaintext passwords in memory
+
+  {
+    std::lock_guard<std::mutex> guard(password_cache_mutex_);
+    auto cached_entry = password_cache_.get(plain);
+    if (cached_entry && *cached_entry == hash) {
+      return true;
+    }
+  }
+
+  // bcrypt_checkpw has stringcmp semantics
+  bool correct = util::bcrypt_checkpw(plain.c_str(), hash.c_str()) == 0;
+  if (correct) {
+    {
+      std::lock_guard<std::mutex> guard(password_cache_mutex_);
+      password_cache_.insert(plain, hash);
```

Comment:
Looks like if another thread managed to insert this mapping before this thread got here, it works just fine because boost's LRU cache seems to treat the "insert" as a no-op if the key already exists (and that's OK for our purposes).

Reply:
I can think of a couple race conditions that might happen if we change the password to be the same, since bcrypt might generate a different hash with a new salt. I'm going to change this to explicitly replace any duplicate entries. Thanks for catching this!